src/Thruway/Procedure.php

```diff
@@ -249,44 +256,93 @@ public function processCall(Session $session, Call $call)
     public function processQueue()
     {
         if (!$this->getAllowMultipleRegistrations()) {
-            throw new \Exception("queuing only allowed when there are multiple registrations");
+            throw new \Exception("Queuing only allowed when there are multiple registrations");
         }
 
         // find the best candidate
         while ($this->callQueue->count() > 0) {
-            $congestion = true;
-
-            /* @var $bestRegistration \Thruway\Registration */
-            $bestRegistration = $this->registrations[0];
-            /* @var $registration \Thruway\Registration */
-            foreach ($this->registrations as $registration) {
-                if ($registration->getSession()->getPendingCallCount() == 0) {
-                    $bestRegistration = $registration;
-                    $congestion       = false;
-                    break;
-                }
-                if ($registration->getSession()->getPendingCallCount() <
-                    $bestRegistration->getSession()->getPendingCallCount()
-                ) {
-                    $bestRegistration = $registration;
-                }
+            $registration = NULL;
+
+            if (strcasecmp($this->getInvokeType(), Registration::FIRST_REGISTRATION) === 0) {
+                $registration = $this->getNextFirstRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::LAST_REGISTRATION) === 0) {
+                $registration = $this->getNextLastRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::RANDOM_REGISTRATION) === 0) {
+                $registration = $this->getNextRandomRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::ROUNDROBIN_REGISTRATION) === 0) {
+                $registration = $this->getNextRoundRobinRegistration();
+            } else if (strcasecmp($this->getInvokeType(), Registration::THRUWAY_REGISTRATION) === 0) {
+                $registration = $this->getNextThruwayRegistration();
             }
 
-            if ($congestion) {
-                // there is congestion
-                $bestRegistration->getSession()->getRealm()->publishMeta('thruway.metaevent.procedure.congestion',
-                    [
-                        ["name" => $this->getProcedureName()]
-                    ]
-                );
+            if ($registration !== NULL) {
+                $call = $this->callQueue->dequeue();
+                $registration->processCall($call);
+            }
```

Comment:
The `$registration` will be `NULL` when the system cannot find a free session to pass the call to. Therefore, in the event that all the sessions of a procedure have blocked (never returning anything), then this will result in an endless loop.

However, this will only happen if the procedure registration `invoke` type is **`thruway`**.

For all other invoke types, there will always be a session, regardless of the fact whether they are free or not.